qi-doc/scribblings/principles.scrbl

```diff
@@ -59,6 +75,171 @@ Consider this example:
+
+In the above example, @racket[filter] and @racket[map] are obviously ordered by @racket[~>] in this way, so that @racket[(filter my-odd?)] is upstream of @racket[(map my-sqr)]. But it's not so obvious how @racket[my-odd?] and @racket[my-sqr] should be treated. These are employed "internally" by the higher-order flows @racket[filter] and @racket[map], and are not directly ordered by the @racket[~>] form. Should @racket[my-odd?] be considered to be upstream of @racket[my-sqr] here?
+
+This is where the distinction between flows and flow invocations comes into play. In fact, not all invocations of @racket[my-odd?] are upstream of any particular invocation of @racket[my-sqr]. Rather, specific invocations of @racket[my-sqr] that use values computed by individual invocations of @racket[my-odd?] are downstream of those invocations, and notably, these invocations involve the individual elements of the input list rather than the entire list, so that the computational dependency expressed by this relation is as fine-grained as possible.
+
+For instance, for an input list @racket[(list 1 2 3)], @racket[(my-odd? 1)] is @tech{upstream} of @racket[(my-sqr 1)], and likewise, @racket[(my-odd? 3)] is @tech{upstream} of @racket[(my-sqr 3)], but @racket[(my-odd? 3)] is not upstream of @racket[(my-sqr 1)], and @racket[(my-odd? 2)] isn't upstream of anything.
+
+That brings us to the guarantee that Qi provides in this case (and in general).
+
+@subsection{Qi's Guarantee on Effects}
+
+@definition["Well-ordering"]{For @tech{flow} invocations @${f} and @${g} and corresponding effects @${ε(f)} and @${ε(g)},
+
+@$${f \lt g \implies \epsilon(f) \lt \epsilon(g)}
+
+where @${<} on the left denotes the relation of being upstream, and @${<} on the right denotes one effect happening before another. Such effects are said to be @emph{well-ordered}.
+}
+
+Well-ordering is defined in relation to a source program encoding the intended meaning of the flow, which serves as the point of reference for program translations. Qi guarantees that effects will remain well-ordered through any such translations of the source program that are undertaken during @seclink["It_s_Languages_All_the_Way_Down"]{optimization}. As we will soon see, this guarantee assumes, and prescribes, that effects employed in flows be @tech{local}.
+
+@definition["Effect locality"]{@tech{Effects} in a flow F are said to be @deftech{local} if the @tech{output} of F is invariant under all @tech{well-orderings} of effects. Specifically, if a @techlink[#:key "well-ordering"]{well-ordered} program translation causes a program to produce different @tech{output}, then the program contains @deftech{nonlocal} effects.}
```

Comment:
The word "local" for this property is a little confusing to me, because the effects may in fact impact each other at a great syntactic or temporal distance! It's just that they're restricted to be related by the dependency structure of the flows.

Reply:
I don't see an explicit definition of what it means to be a well-ordered program transformation, as opposed to effects in a program being well-ordered. At the same time, bringing in the issue of program transformations at all and what properties they need to maintain seems like it should be in separate developer documentation.